mainwindow.cpp

```diff
@@ -0,0 +1,334 @@
+        }
+    }
+}
+
+void MainWindow::SelectLetters_Slot()
+{
+    SelectFilesAndRefreshLabels();
+}
+
+bool MainWindow::isValidEmail(const QString &email)
+{
+    static QRegularExpression s_email_regex("^[\\w\\.-]+@[\\w\\.-]+\\.\\w+$");
+
+    QRegularExpressionMatch match = s_email_regex.match(email);
+    return match.hasMatch();
+}
+
+
+bool MainWindow::CheckEmails(const QLineEdit* Container)
+{
+    static QRegularExpression s_split_regex("\\s*,\\s*|\\s+");
+
+    QString recipients = Container->text();
+    QStringList email_list = recipients.split(s_split_regex, Qt::SkipEmptyParts);
+    if (email_list.isEmpty())
+    {
+        QMessageBox::critical(this, "No recipients!", "You haven't written any recipients!");
+        return false;
+    }
+
+    QStringList invalid_emails;
+
+    for (const QString &email : email_list)
+    {
+        if (!isValidEmail(email))
+        {
+            invalid_emails += email;
+        }
+    }
+
+    if (!invalid_emails.isEmpty())
+    {
+        QString msg = R"*(Invalid emails: ")*";
+        msg += invalid_emails.join("; ");
+        msg += R"*("!)*";
+
+        QMessageBox::critical(this, "Invalid email detected!", msg);
+        return false;
+    }
+    return true;
+}
+
+void MainWindow::SpawnNewHistoryUnit(const LetterStruct& Letter)
+{
+    QVBoxLayout* layout = qobject_cast<QVBoxLayout*>(ui->MailHistoryScrollArea->layout());
+    MailHistoryUnit* new_history_unit = nullptr;
+
+    if (layout)
+    {
+        new_history_unit = new MailHistoryUnit(Letter);
+        layout->setAlignment(Qt::AlignHCenter | Qt::AlignTop);
+        layout->insertWidget(0, new_history_unit);
+    }
+
+    connect(new_history_unit, SIGNAL(OnMouseReleased(QVector<LetterStruct>)), this, SLOT(HistoryWidgetClicked(QVector<LetterStruct>)));
+}
+
+void MainWindow::SpawnNewHistoryUnit(const QVector<LetterStruct>& Letters)
+{
+    QVBoxLayout* layout = qobject_cast<QVBoxLayout*>(ui->MailHistoryScrollArea->layout());
+    MailHistoryUnit* new_history_unit = new MailHistoryUnit(Letters);
+
+    if (layout)
+    {
+        new_history_unit = new MailHistoryUnit(Letters);
+        layout->setAlignment(Qt::AlignHCenter | Qt::AlignTop);
+        layout->insertWidget(0, new_history_unit);
+    }
+
+    connect(new_history_unit, SIGNAL(OnMouseReleased(QVector<LetterStruct>)), this, SLOT(HistoryWidgetClicked(QVector<LetterStruct>)));
+}
+
+void MainWindow::PopulateMailsHistory()
+{
+    QDir dir(m_temp_file_path);
+
+    if (!dir.exists())
+    {
+        qDebug() << "Directory does not exist:" << m_temp_file_path;
+    }
+
+    dir.setSorting(QDir::Time);
+    dir.setFilter(QDir::Files);
+    QStringList files = dir.entryList();
+
+    for (const QString &fileName : files)
+    {
+        qDebug() << fileName;
+        QVector<LetterStruct> letters = ReadLettersFromFile(m_temp_file_path + fileName);
+        if (!letters.isEmpty()) SpawnNewHistoryUnit(letters);
+    }
+}
+
+void MainWindow::SendLetter(const LetterStruct& Letter)
```

Comment:
where did the function body go?

Reply:
It's an empty shell for a future function yet to be made by a client team